llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp

```diff
@@ -2454,8 +2454,10 @@ bool SIInsertWaitcnts::isPreheaderToFlush(
 }
 
 bool SIInsertWaitcnts::isVMEMOrFlatVMEM(const MachineInstr &MI) const {
-  return SIInstrInfo::isVMEM(MI) ||
-         (SIInstrInfo::isFLAT(MI) && mayAccessVMEMThroughFlat(MI));
+  if (SIInstrInfo::isFLAT(MI))
+    return mayAccessVMEMThroughFlat(MI);
+  return SIInstrInfo::isMUBUF(MI) || SIInstrInfo::isMTBUF(MI) ||
+         SIInstrInfo::isImage(MI);
```

Comment:
Simpler:
```suggestion
  return SIInstrInfo::isVMEM(MI);
```

Reply:
I think Matt wanted me to list them separately: https://github.com/llvm/llvm-project/pull/137148#discussion_r2058039211

Reply:
That's not how I interpreted Matt's comment, but it's not important anyway.

Reply:
I mean the special case handling where it's looking at the memory operands. mayAccessVMEMThroughFlat can only improve a very niche case. To handle that, treat isFLAT as the special case and then the rest doesn't matter, it's the vmem except flat

Reply:
Alright, thanks for the clarification, fixed it :)